Avoid pcluster managing of custom Slurm partitions

hanwen-pcluste opened the pull request:
### Description of changes

Custom Slurm partitions will no longer be managed by pcluster clustermgtd.

Below is a list of behavior change, which may be not exhaustive
1. Custom partitions 
Old behavior: custom partitions had been being managed by clustermgtd. 
New behavior: custom partitions will not be managed by clustermgtd. 
    1. pcluster update-update-compute-fleet
Old behavior: the command had been changing status to all partitions. 
New behavior: the command will change status only to pcluster partitions. 
    1. when custom partitions are set to inactive
Old behavior: clustermgtd had been terminating nodes belonging to the partitions. 
New behavior: clustermgtd won't do anything.
    1. protected mode
Old behavior: Any compute resource had been able to trigger protected mode, regardless of the partitions are pcluster/custom. When protected mode was triggered, all partitions containing the compute resource were set to inactive, regardless of the partitions are managed/custom. 
New behavior: Only compute resource belonging to pcluster partitions will be able to trigger protected mode. When protected mode is triggered, only pcluster partitions containing the compute resource will be set to inactive.
3. Custom nodes with name format: `<queue name>-st/dy-<compute resource name>-\d`
Old behavior: the nodes had been managed by clustermgtd. 
New behavior: if the nodes are in pcluster partition, clustermgtd will manage it. Otherwise, clustermgtd will not manage it.
4. Custom nodes with other name formats Behavior is not changed. the nodes weren't and won't be managed by clustermgtd

Also, some small refactoring was done to simplify the logic in some of the node functions and methods and to remove unused functions.

### Tests
Wrote new integration tests https://github.com/aws/aws-parallelcluster/pull/5442 and unit tests
The following integration tests have been passed

```
{%- import 'common.jinja2' as common with context -%}
---
test-suites:
  schedulers:
    test_slurm.py::test_slurm:
      dimensions:
        - regions: [ "eu-central-1" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "ubuntu2004" ]
          schedulers: [ "slurm" ]
    test_slurm.py::test_slurm_pmix: # TODO: include in main test_slurm to reduce number of created clusters
      dimensions:
        - regions: [ "ap-south-1" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "centos7" ]
          schedulers: [ "slurm" ]
        - regions: [ "ap-northeast-1" ]
          instances: {{ common.INSTANCES_DEFAULT_ARM }}
          oss: [ "rhel8" ]
          schedulers: [ "slurm" ]
    test_slurm.py::test_slurm_scaling:
      dimensions:
        - regions: [ "use2-az2" ]  # do not move, unless instance type support is moved as well
          instances: [ {{ common.instance("instance_type_1") }} ]
          oss: [ "alinux2" ]
          schedulers: [ "slurm" ]
    test_slurm.py::test_error_handling:
      dimensions:
        - regions: [ "ca-central-1" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "alinux2" ]
          schedulers: [ "slurm" ]
    test_slurm.py::test_slurm_protected_mode:
      dimensions:
        - regions: [ "ca-central-1" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "rhel8" ]
          schedulers: [ "slurm" ]
    test_slurm.py::test_slurm_protected_mode_on_cluster_create:
      dimensions:
        - regions: [ "ap-east-1" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "centos7" ]
          schedulers: [ "slurm" ]
    test_slurm.py::test_slurm_memory_based_scheduling:
      dimensions:
        - regions: [ "ap-east-1" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "ubuntu2004" ]
          schedulers: [ "slurm" ]
    test_slurm.py::test_scontrol_reboot:
      dimensions:
        - regions: [ "us-east-1" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "ubuntu2004" ]
          schedulers: [ "slurm" ]
    test_slurm.py::test_scontrol_reboot_ec2_health_checks:
      dimensions:
        - regions: [ "us-east-2" ]
          instances: [ "t2.medium" ]
          oss: [ "ubuntu2004" ]
          schedulers: [ "slurm" ]
    test_slurm.py::test_slurm_overrides:
      dimensions:
        - regions: [ "me-south-1" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "ubuntu2004" ]
          schedulers: [ "slurm" ]
    test_slurm.py::test_scontrol_update_nodelist_sorting:
      dimensions:
        - regions: [ "ca-central-1" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "alinux2" ]
          schedulers: [ "slurm" ]
    test_slurm_accounting.py::test_slurm_accounting:
      dimensions:
        - regions: [ "ap-south-1" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "ubuntu2004" ]
          schedulers: [ "slurm" ]
    test_slurm_accounting.py::test_slurm_accounting_disabled_to_enabled_update:
      dimensions:
        - regions: [ "us-west-2" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "centos7" ]
          schedulers: [ "slurm" ]
    test_slurm.py::test_slurm_reconfigure_race_condition:
      dimensions:
        - regions: [ "af-south-1" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "alinux2" ]
          schedulers: [ "slurm" ]
    test_slurm.py::test_slurm_custom_partitions:
      dimensions:
        - regions: [ "us-east-1" ]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: [ "alinux2" ]
          schedulers: [ "slurm" ]
```

Some unit tests were added to cover the new logic.

### References
This PR depends on https://github.com/aws/aws-parallelcluster-cookbook/pull/2323

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jdeamicis committed:
Avoid pcluster managing of custom Slurm partitions

Custom Slurm partitions will no longer be managed by pcluster clustermgtd.

Below is a list of behavior change, which may be not exhaustive
1. Custom partitions
Old behavior: custom partitions had been being managed by clustermgtd.
New behavior: custom partitions will not be managed by clustermgtd.
1.1 pcluster update-update-compute-fleet
Old behavior: the command had been changing status to all partitions.
New behavior: the command will change status only to pcluster partitions.
1.2 when custom partitions are set to inactive
Old behavior: clustermgtd had been terminating nodes belonging to the partitions.
New behavior: clustermgtd won't do anything.
1.3 protected mode
Old behavior: Any compute resource had been able to trigger protected mode, regardless of the partitions are pcluster/custom. When protected mode was triggered, all partitions containing the compute resource were set to inactive, regardless of the partitions are managed/custom.
New behavior: Only compute resource belonging to pcluster partitions will be able to trigger protected mode. When protected mode is triggered, only pcluster partitions containing the compute resource will be set to inactive.
2. Custom nodes with name format: <queue name>-st/dy-<compute resource name>-\d
Old behavior: the nodes had been managed by clustermgtd.
New behavior: if the nodes are in pcluster partition, clustermgtd will manage it. Otherwise, clustermgtd will not manage it.
3. Custom nodes with other name formats
Behavior is not changed. the nodes weren't and won't be managed by clustermgtd

jdeamicis commented on src/slurm_plugin/clustermgtd.py:
```diff
@@ -739,14 +745,14 @@ def _increase_partitions_protected_failure_count(self, bootstrap_failure_nodes):
-                    self._partitions_protected_failure_count_map[p] = {}
-                    self._partitions_protected_failure_count_map[p][compute_resource] = 1
+            queue_name = node.queue_name
+            if queue_name in self._partitions_protected_failure_count_map:
+                self._partitions_protected_failure_count_map[queue_name][compute_resource] = (
+                    self._partitions_protected_failure_count_map[queue_name].get(compute_resource, 0) + 1
+                )
+            else:
+                self._partitions_protected_failure_count_map[queue_name] = {}
+                self._partitions_protected_failure_count_map[queue_name][compute_resource] = 1
```
Nitpick: can't we use the `dict.setdefault()` method the same way we use it below in line 1065?

jdeamicis reviewed (changes requested):
There are several new lines that are not covered by unit tests. We must address this before merging.

jdeamicis committed:
Add test to load partition-nodelist mapping

jdeamicis committed:
Add TestPartitionNodelistMapping test class

Add unit tests for the whole PartitionNodelistMapping singleton
class.

jdeamicis commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -249,7 +293,16 @@ def get_nodes_info(nodes="", command_timeout=DEFAULT_GET_INFO_COMMAND_TIMEOUT):
+    TODO: we can consider building a filter to be used in case a nodes argument is passed, in order to exclude nodes
+     not managed by ParallelCluster.
     """
+    if nodes == "":
+        nodes = _get_all_partition_nodes(",".join(PartitionNodelistMapping.instance().get_partitions()))
```
Do we still want this with the logic we chose to implement (i.e. _allowing_ users to add nodes in PC-managed partitions)?

jdeamicis committed:
Add unit tests for get_nodes_info()

Add functional unit tests to verify that the correct nodelists are passed
to the Slurm commands when retrieving compute node information.

Add typing hints to the PartitionNodelistMapping class methods.

jdeamicis committed:
[Refactoring] remove unused `_get_partition_nodes()` function

Remove unused function from the node code. As a consequence,
simplify the body of `get_partition_info()`.

jdeamicis committed:
Add typing for test_grep_partition_filter test

jdeamicis committed:
Add typing for get_nodes_info() function

Add TODO to possibly remove resume_powering_down_nodes() function.
This would also allow to remove the _get_slurm_nodes() function.

jdeamicis commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -70,6 +72,48 @@
+
+    def get_partition_nodelist_mapping(self) -> Dict[str, str]:
+        """Retrieve partition-nodelist mapping from JSON file."""
+        if not self.partition_nodelist_mapping:
+            partition_nodelist_json = os.path.join(
```
I don't understand why Codecov complains about this: this should be covered by the new `test_get_partition_nodelist_mapping` defined below (see `test_slurm_commands.py`).

hanwen-pcluste replied:
~~Because we mocked it in the unit tests. I think we can leave this as is~~

jdeamicis replied:
We are not really mocking the reading of the file, we are only mocking the value of the global variable `SLURM_CONF_DIR` to be under the test data directory of pytest.

I believe Codecov is getting crazy here.

jdeamicis commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -70,6 +72,48 @@
+                SLURM_CONF_DIR,
+                "pcluster/parallelcluster_partition_nodelist_mapping.json",
+            )
+            with open(partition_nodelist_json, "r", encoding="utf-8") as file:
+                self.partition_nodelist_mapping = json.load(file)
+        return self.partition_nodelist_mapping
```
Same thing here (see comment above).

hanwen-pcluste replied:
Same answer

jdeamicis replied:
We are not really mocking the reading of the file, we are only mocking the value of the global variable `SLURM_CONF_DIR` to be under the test data directory of pytest.

I believe Codecov is getting crazy here.

jdeamicis commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -261,9 +314,18 @@ def get_nodes_info(nodes="", command_timeout=DEFAULT_GET_INFO_COMMAND_TIMEOUT):
+
+    This function considers only partitions managed by ParallelCluster.
+    """
+    partitions = list(PartitionNodelistMapping.instance().get_partitions())
+    grep_filter = _get_partition_grep_filter(partitions)
+    show_partition_info_command = (
```
It makes little sense to test this if we have to mock the result of the actual Slurm command.

The `PartitionNodelistMapping.get_partitions()` part is covered by the new test class `TestPartitionNodelistMapping` tests.

Another unit test was created for the `get_partition_grep_filter` function.

The third line is a simple Python f-string.

hanwen-pcluste reviewed (commented):
LGTM. Because I cannot approve my own PR. @jdeamicis , feel free to approve the PR

hanwen-cluster committed:
Modify changelog

jdeamicis committed:
Rename get_partition_info() to get_partitions_info()

Renamed caller function _get_partition_info_with_retry() accordingly.

Add typing hints to get_partitions_info() and
_get_partitions_info_with_retry().

jdeamicis committed:
Improve robustness of _get_partition_grep_filter()

Make grep filter function compatible with empty lists of partitions
(it should not be possible as of now, but it may be useful in the
future).
